keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,750 @@
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+reviewers:
+  - "@klueska"
+  - "@derekwaynecarr"
+approvers:
+  - TBD
+  - TBD
+editor: TBD
+creation-date: 2020-02-03
+last-updated: TBD
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Memory Manager
+
+<!-- 
+_Authors:_
+
+- @bg-chun - Byonggon Chun &lt;bg.chun@samsung.com&gt;
+- @cezaryzukowski - Cezary Zukowski &lt;c.zukowski@samsung.com&gt;
+- @alukiano - Artyom Lukianov &lt;alukiano@redhat.com&gt;
+-->
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1 : High-performance packet processing with DPDK](#story-1--high-performance-packet-processing-with-dpdk)
+    - [Story 2 : Databases](#story-2--databases)
+  - [Notes/Constraints/Caveats (optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Memory Map](#memory-map)
+    - [Memory Maps at start-up (with examples)](#memory-maps-at-start-up-with-examples)
+    - [Memory Maps at runtime (with examples)](#memory-maps-at-runtime-with-examples)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+The *Memory Manager* is a new component in *kubelet* ecosystem proposed to enable the feature of guaranteed memory (and hugepages) allocation for pods in Guaranteed QoS class. The feature offers a couple of allocation strategies. The first one, the *single-NUMA* strategy, is intended for high-performance and performance-sensitive applications. Relevant user stories are illustrated in here for the *single-NUMA* strategy. The second one, the *multi-NUMA* strategy, complements the overall design while it overcomes the situation that cannot be managed with the *single-NUMA* strategy. Namely, whenever the amount of memory demanded by a pod is in the excess exceeds of a single NUMA node capacity, the guaranteed memory is provisioned across multiple NUMA nodes with the *multi-NUMA* strategy.  
```

Comment:
Do we really need new component? Would it make sense to combine memory and CPU allocations in one component? Either by extending CPU manager or by creating Native Resource manager that would handle both CPU and Memory allocations?

Reply:
1. I agree it tightly coupled, but it still different components under the operating system, and in theory it possible that you want to use only the CPU manager or memory manager, but not both.
2. I think it easier to develop the stuff when you have functionality separated between managers, so you do not really need to be afraid that you will break the backward compatibility of the CPU manager.
3. The memory manager will be an alpha feature, so it will need to introduce additional conditional checks under the CPU manager, which also will not make the code more readable.
4. It has a different structure under the state file and it has a different logic hints generation, compose CPU manager and memory can again make code more complex.
5. We do not need the reconcile loop under the memory manager.
6. We have different types of memory, so I think it just natural to work with all of them under the single manager.

I like the idea of Native Resource manager, but in this case, it should be done as step two and only in the case when you really want to introduce more complex hint calculations that will see CPU and memory as a single unit(all different permutations of NUMA nodes and CPU's).

Reply:
> in theory it possible that you want to use only the CPU manager or memory manager, but not both.

Do you have any real life examples of this? Even if they exists that can be an option for the combined memory and CPU manager.

>  think it easier to develop the stuff when you have functionality separated between managers, so you do not really need to be afraid that you will break the backward compatibility of the CPU manager.

> The memory manager will be an alpha feature, so it will need to introduce additional conditional checks under the CPU manager, which also will not make the code more readable.

Native Resource manager would eliminate both these concerns I believe.

> compose CPU manager and memory can again make code more complex.

Agree. However, it makes sense to discuss this further, find out proc and cons of each approach and select the best of two in my opinion.

> We have different types of memory, so I think it just natural to work with all of them under the single manager.

HugePages and Memory are quite different resources from usage point of view. Following your logic why not to separate them too?

Reply:
> Do you have any real life examples of this? Even if they exists that can be an option for the combined memory and CPU manager.

The only case that I can imagine, that if the kernel will have some better heuristics for the CPU or memory allocation in the future, a user will want to leave the managing of one of the resources to the kernel without setting the pinning(I know that it possible to use NUMA auto-balancing feature, I also know that this feature disabled under RT workflows).
I believe it reasonable to ask why do we have cpuset separation for the memory and CPU's under the kernel.

> Native Resource manager would eliminate both these concerns I believe.

Can you please explain how it will eliminate these concerns?

> Agree. However, it makes sense to discuss this further, find out proc and cons of each approach and select the best of two in my opinion.

Can you please start a document that will describe pros of the `Native Resource manager` and we will add our thoughts under it?

> HugePages and Memory are quite different resources from usage point of view. Following your logic why not to separate them too?

Because you can not pin memory and hugepages of the same process to different NUMA nodes when you can do it for the CPUs and memory.

Reply:
What would be the situtation when you would use memory pinning without explicit knowledge about which CPU's the container is going to get pinned to ? And how would you use memory pinning then ? IOW, is there any sane/valid situation when you would use memory pinning but not CPU pinning ?
￼
You might want to pin CPU's only without memory pinning, but I can't come up with any reason to do the other way around.

So it looks to me as
  1. You never pin memory without pinning CPUs
  2. To pin memory, you must know which CPU's you're going to get pinned to
=> It looks to me as if memory pinning would be more like an extra 'feature/flag' of CPU-pinning, one which you might or might not be using, but always in addition to CPU pinning...

Reply:
I am thinking it is a more architectural decision. We do not prevent from using the memory manager without the CPU manager, but it is your choice like we do not prevent to use CPU manage and device manager without the topology manager, I can say it also can be very inefficient. We have requested to provide some best-practices document, where we can specify that the usage of the memory manager without the CPU manager is inefficient, and provide kubelet configuration samples for different types of systems. But again from my perspective developing and maintain the memory manager as the separate component is much easier than trying to add this functionality under the CPU manager, when the CPU manager developed without concern about the memory.

Reply:
I believe that configuring one component is much easier than 2 from user point of view. Considering than pinning memory is never done without pinning CPUs I think we should discuss this topic again. From both architectural and usability points of view separating memory management from CPU management still looks questionable to me.

I'd propose to discuss this on the next Sig-Node or Resource management WG meeting.

Reply:
I think from the usability perspective it no difference because you need memory manager flags anyway setting two flags because you have two separate cases:
1. A user wants only CPU pinning
2. A user wants CPU pinning and memory pinning(for this one we will need the flag)

And you still need to provide `reserved-memory` flag, because `system-reserved` and `kube-reserved`, not NUMA aware(the second flag).

Regarding architectural design, I think it much easier to support the memory manager as a separate component from the maintainability perspective. We avoid introducing a new code under the CPU manager that does not really relate to the CPU and that can bring complexity increasing of the existing code under the CPU manager.